nixos/modules/tasks/filesystems/zfs.nix

```diff
@@ -78,6 +85,80 @@ let
     }
   '';
 
+  getPoolFilesystems = pool:
+    filter (x: x.fsType == "zfs" && (fsToPool x) == pool) config.system.build.fileSystems;
+
+  getPoolMounts = prefix: pool:
+    let
+      mountPoint = fs: escapeSystemdPath (prefix + fs.mountPoint);
+    in
+      map (x: "${mountPoint x}.mount") (getPoolFilesystems pool);
+
+  createImportService = { pool, systemd, force, prefix ? "" }:
+    nameValuePair "zfs-import-${pool}" {
+      description = "Import ZFS pool \"${pool}\"";
+      # we need systemd-udev-settle until https://github.com/zfsonlinux/zfs/pull/4943 is merged
+      requires = [ "systemd-udev-settle.service" ];
+      after = [
+        "systemd-udev-settle.service"
+        "systemd-modules-load.service"
+        "systemd-ask-password-console.service"
+      ];
+      wantedBy = (getPoolMounts prefix pool) ++ [ "local-fs.target" ];
+      before = (getPoolMounts prefix pool) ++ [ "local-fs.target" ];
+      unitConfig = {
+        DefaultDependencies = "no";
+      };
+      serviceConfig = {
+        Type = "oneshot";
+        RemainAfterExit = true;
+      };
+      environment.ZFS_FORCE = optionalString force "-f";
+      script = (importLib {
+        # See comments at importLib definition.
+        zpoolCmd = "${cfgZfs.package}/sbin/zpool";
+        awkCmd = "${pkgs.gawk}/bin/awk";
+        inherit cfgZfs;
+      }) + ''
+        poolImported "${pool}" && exit
+        echo -n "importing ZFS pool \"${pool}\"..."
+        # Loop across the import until it succeeds, because the devices needed may not be discovered yet.
```

Comment:
Why is this needed? I thought it's what udev-settle was for?

Reply:
udev-settle ensures that all pending udev events have been processed. It does _not_ trigger any logic to wait for all hardware probing to finish, etc. On PCs this rarely makes an issue, but when I wrote this, it was to deal with a server-class motherboard that wouldn't finish probing HDDs (and passing them to the OS) until _a minute and a half after the kernel had booted_.

This is not unusual for servers, as I understand it.

In other words:

- This code is untestable on any hardware you're likely to own. Or which I own, for that matter. The C2750 is far too slow to use now.
- If we get it wrong, then a rare few users will be very sad. These are the exact same users most likely to care about reliability.